CI: Try installing doxygen from binaries.

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits try to add a way of installing Doxygen from its latest binaries since apt only has 1.9.8 available causing the strange error we found during CI in #15306.

This new approach has hardcoded version 1.15.0 of doxygen as the default instalation for linux based systems.

### Details and comments
Based on an error found in #15306

raynelfss committed:
CI: Try installing doxygen from binaries.

raynelfss committed:
Fix: Handle receipt of github token.

raynelfss committed:
Fix: Close script after error.

raynelfss committed:
FIx: Try using as env variable.

raynelfss committed:
Fix: Wrong line

raynelfss committed:
Fix: Try single bracket.

raynelfss committed:
Fix: Try again with correct substitution

raynelfss committed:
Fix: Find correct folder

raynelfss committed:
FIx: Try to cd correctly

raynelfss committed:
Fix: Erroneous comment of previous commands

raynelfss committed:
Fix: Try saving directory in a variable.

raynelfss committed:
Fix: Remove the ./configure part.

raynelfss committed:
Fix: Try with sudo make

mtreinish commented on tools/install_ubuntu_docs_dependencies.sh:
```diff
@@ -8,4 +8,20 @@ python -m pip install --upgrade pip setuptools wheel
+cd ./doxygen
+./configure
+make
+sudo make install
```
Doxygen publishes precompiled binaries, this is building it from source? Why not just download: `https://github.com/doxygen/doxygen/releases/download/Release_1_15_0/doxygen-1.15.0.linux.bin.tar.gz` and run untar it and run from `doxygen-1.15.0/bin/doxygen`? Or just run make install to install it system wide

We also don't need to always install the latest we can just use a fixed version that works.

raynelfss replied:
That is exactly what it does, I just missed a couple of steps along the way. But this file is the one including the linux binaries, I just renamed it to make it easier to decompress. So I just run `sudo make` and `sudo make install` to install the files. Unless I just need `sudo make install`.

raynelfss commented:
As of 23b8ca2, doxygen is working with the latest version. However, the fixes @jakelishman applied in #15306 are not applied and so it complains still.

jakelishman commented on tools/install_ubuntu_docs_dependencies.sh:
```diff
@@ -2,10 +2,30 @@
 
-set -e
+if [ $GITHUB_TOKEN ]; then
+    set -e
```
You might want something more like
```bash
set -e

if [ -z "$GITHUB_TOKEN ]; then
    echo "No github token" >&2
    exit 1;
fi

# ... rest of the script
```

to avoid everything being indented.

raynelfss replied:
Added in 40064654f019d7a95774c978a3245994a678b543

jakelishman commented on tools/install_ubuntu_docs_dependencies.sh:
```diff
@@ -2,10 +2,30 @@
+    curl -H "Authorization: token $GITHUB_TOKEN" \
+        https://api.github.com/repos/doxygen/doxygen/releases/latest | \
+        jq '.["assets"][2]["browser_download_url"]' | \
+        xargs -I {} wget {} -O ./doxygen.tar.gz
```
Using `.["assets"][2]["browser_download_url"]` is assuming that the order the items come out in the release will always be the same, which is fragile.  But we probably don't want to attempt to get the "latest" anyway - it's more reliable to stick with a known-good version, at which point we can just hardcode the link.

raynelfss replied:
Hardcoded 1.15.0 in 40064654f019d7a95774c978a3245994a678b543

jakelishman commented on tools/install_ubuntu_docs_dependencies.sh:
```diff
@@ -2,10 +2,30 @@
+    cd $DOXYGEN_DIR
+    # Run the remainder of the setup process
+    sudo make
+    sudo make install
```
The Makefile of the binary distribution only has an `install` rule, so this is running `sudo make install` twice (though that's harmless). If we use a hard-coded Doxygen version, we won't need to futz around with the searching, though if we did leave it in place, you're already relying on there exactly one match to the glob, so you could `shopt failglob` (for safety) and then just `cd ./doxygen*` directly.

raynelfss replied:
Since the filename will match if we hard-code it, I made the cd use a hard-coded name in 40064654f019d7a95774c978a3245994a678b543

jakelishman reviewed (commented):
Thanks for looking at this Ray.  If you wanted to merge this PR separately, you could have it disable warnings-are-errors, then turn them back on in your other if you liked.  It's also ok just to join this PR into the other one too, though.

raynelfss committed:
Fix: Review comments
- Hardcode binaries from script
- Remove unnecessary indentation in favor of a quick exit.
- Remove duplicate calls to make.

raynelfss commented:
> If you wanted to merge this PR separately, you could have it disable warnings-are-errors, then turn them back on in your other if you liked

For this PR, I'd rather solve the warnings being pointed out here and merge them with it. It's a more conservative route but it's less risky. The other failures fixed in previous commits in the other PR should merge with that.

raynelfss committed:
Fix inline-code typos

Co-authored-by: Jake Lishman <jake@binhbar.com>

raynelfss commented:
Now I wonder if we will need to add some documentation as to how we can update the version of doxygen if the time comes. Is there a procedure for CI-based updates?

jakelishman commented on tools/install_ubuntu_docs_dependencies.sh:
```diff
@@ -3,9 +3,24 @@
 # Prepare an Ubuntu CI machine for running 'tox -e docs'.  Assumes that Python is available.
 
 set -e
+if [ -z $GITHUB_TOKEN ]; then
+    echo "No github token provided"
+    exit 1
+fi
```
I suppose technically we could just fall back to doing the `wget` with no authentication in this case, but it doesn't matter.

jakelishman reviewed (approved):
Thanks Ray, this looks like much cleaner now.  Fwiw, I was able to reproduce the bugs using Doxygen 1.9.8 built from source (what was in the Ubuntu repos), but not 1.9.4 (in the Debian repos my Raspberry Pi was pulling from), so it does seem like it was ephemeral.

For updating the version: the bash commands you've written here are standard stuff, and it's fairly clear how to update imo, so I wouldn't worry about it.
